Adding preflight skeleton

ajCameron opened the pull request:
Contains the programmatic parts of https://github.com/mewbotorg/mewbot/pull/175 - which is deprecated by this pull request.

 - preflight is intended to be a single script you run before submitting a pull request
 - it runs reuse, linters, then the tests in order and reports on the result
 - reuse has been added to tools as a script - and to mewbot.tools as a runnable module
 - terminal module has been added to tools with some helpful delineation functions

ajCameron committed:
Adding preflight skeleton

 - preflight is intended to be a single script you run before submitting a pull request
 - it runs reuse, linters, then the tests in order and reports on the result
 - reuse has been added to tools as a script - and to mewbot.tools as a runnable module
 - terminal module has been added to tools with some helpful delineation functions

ajCameron committed:
Removing some sonarcloud code smells

 - merging some strings which had gotten split up
 - replacing pass in for with ellipsis - which may help indicate all we need is to run the iterator

exachixkitsune commented on src/mewbot/tools/lint.py:
```diff
@@ -38,14 +39,24 @@ class LintToolchain(ToolChain):
+            yield from self.lint_pydocstyle()
+
+        if self.is_ci and self.fail_ci_run:
+            print(f"Some of the tools reports having silently failed!\n{self.run_success}")
```
What's stopping this from appearing if something hasn't silently failed? I cannot see where that would occur.

javajawa replied:
This has been removed :)

exachixkitsune reviewed (dismissed):
Only one minor comment; looks good to me!

javajawa reviewed (changes requested):
Loving the pre-flight part, I'm going to have a little play around combining the `CommandDeliminator` with the existing login in the `_run_utility` function -- the pipelines are looking a little messier: https://github.com/mewbotorg/mewbot/actions/runs/4993044197/jobs/8941766975

javajawa committed:
Consistency Fix: is_ci -> in_ci

javajawa committed:
Mark tools/preflight as executable

javajawa committed:
Move ToolChain.run_utility head to us CommandDelimiter

javajawa committed:
Refactor ReuseRun to ToolChain and rename CommandDeliminator to CommandDelimiter

javajawa committed:
Refactor PreFlight into a ToolChain

ToolChain.success is removed in favour of a property that check each sub-command's status.

javajawa committed:
Change ToolChain summary to a more grid like pattern

The two statuses that can occur are now PASS and FAIL,
and the value is put at the beginning on the line.
The results in the state messages being aligned in the terminal.

javajawa committed:
Teach CommandDelimiter about output modes

 - The max length the header line can be is now 80 (or terminal width)
 - The header line is in 'bold' white (generally grey) to make it easier to spot
 - The CommandDelimiter only outputs one line in terminal mode, keeping the terminal cleaner
 - In CI mode, the CommandDelimiter outputs GitHub action group annotations

javajawa committed:
Add 'reuse lint' to the reuse toolchain

javajawa committed:
Remove redundant CI failure check from linting toolchain

javajawa commented:
I have a series of proposals that are a little too complicated for pull request comments, so I have opened a dependant PR #197 with my thinking.

The first bits are minor fixes/corrections:
 - Consistency Fix: is_ci -> in_ci (not really related to this PR, but it made things easier)
 -  Mark `tools/preflight` as executable (a little `chmod` action)
 - Push the `CommandDeliminator` call down into `ToolChain._run_utility`
   - I toyed around here with the idea of it prefix each output line with the command's name
 - Put a loop call when calling the test runs so preflight actually runs them.

Then come the big changes:
 - Refactor `ReuseRun` to `ReuseToolchain` and later add `reuse lint` as a step.
 - Refactor `PreFlight` to a ToolChain, to take advantage of code reuse.
 - Change the big Report format to looks like 
![2023-05-20_300x226](https://github.com/mewbotorg/mewbot/assets/653482/d3bbf99f-8b37-43b4-a8f4-07cf634bbaea)
 - Change the command delimitor to look like 
![2023-05-20_666x158](https://github.com/mewbotorg/mewbot/assets/653482/a48637ad-9b04-46de-a200-68c0ca722bb5)
 - In CI, it looks like ![2023-05-20_693x467](https://github.com/mewbotorg/mewbot/assets/653482/489a40e0-7d13-404a-a8b0-f8b29097f294)

@ajCameron if you're happy with all that, you can fast-forward this branch to that one, and GitHub will mark 197 as merged without any need for a merge commit :)

javajawa committed:
Move ToolChain to different module

This allows early-bird tools like 'install' and 'path' to run without clint installed

javajawa committed:
Flush stdout when entering a command delimiter
